File: js/moq/src/ietf/object.ts
import type { Reader, Writer } from "../stream";

const SUBGROUP_ID = 0x0; // Must always be layer 0
const STREAM_TYPE = 0x04;
const GROUP_END = 0x03;

/**
 * STREAM_HEADER_SUBGROUP from moq-transport spec.
 * Used for stream-per-group delivery mode.
 */
export class Group {
	static id = STREAM_TYPE;

	subscribeId: bigint;
	trackAlias: bigint;
	groupId: number;
	publisherPriority: number;

	constructor(subscribeId: bigint, trackAlias: bigint, groupId: number, publisherPriority: number) {
		this.subscribeId = subscribeId;
		this.trackAlias = trackAlias;
		this.groupId = groupId;
		this.publisherPriority = publisherPriority;
	}

	async encode(w: Writer): Promise<void> {
		// Stream type is written by the caller
		await w.u62(this.subscribeId);
		await w.u62(this.trackAlias);
		await w.u53(this.groupId);
		await w.u8(SUBGROUP_ID);
		await w.u8(this.publisherPriority);
	}

	static async decode(r: Reader): Promise<Group> {
		const subscribeId = await r.u62();
		const trackAlias = await r.u62();
		const groupId = await r.u53();
		const subgroupId = await r.u53();
		if (subgroupId !== SUBGROUP_ID) {
			throw new Error(`Unsupported subgroup id: ${subgroupId}`);
		}
		const publisherPriority = await r.u8();

		return new Group(subscribeId, trackAlias, groupId, publisherPriority);
	}
}

export class Frame {
	id: number;

	// undefined means end of group
	payload?: Uint8Array;

	constructor(id: number, payload?: Uint8Array) {
		this.id = id;
		this.payload = payload;
	}

	async encode(w: Writer): Promise<void> {
		await w.u53(this.id);

		if (this.payload !== undefined) {
			await w.u53(this.payload.byteLength);

			if (this.payload.byteLength === 0) {
				await w.u8(0); // status = normal
			} else {
				await w.write(this.payload);
			}
		} else {
			await w.u8(0); // length = 0
			await w.u8(GROUP_END);
		}
	}

	static async decode(r: Reader): Promise<Frame> {
		const id = await r.u53();

		const payloadLength = await r.u53();

		if (payloadLength > 0) {
			const payload = await r.read(payloadLength);
			return new Frame(id, payload);
		}

		const status = await r.u53();

		// TODO status === 0 should be an empty frame, but moq-rs seems to be sending it incorrectly on group end.
		if (status === 0 || status === GROUP_END) {
			return new Frame(id);
		}

		throw new Error(`Unsupported object status: ${status}`);
	}
}

/**
 * Helper to read a stream type from a reader.
 */
export async function readStreamType(r: Reader): Promise<void> {
	const streamType = await r.u53();
	if (streamType !== STREAM_TYPE) {
		throw new Error(`Unsupported stream type: ${streamType}`);
	}
}

/**
 * Helper to write a stream type to a writer.
 */
export async function writeStreamType(w: Writer): Promise<void> {
	await w.u53(STREAM_TYPE);
}

File: rs/moq/src/ietf/object.rs
use bytes::Buf;

use crate::coding::{Decode, DecodeError, Encode};

const SUBGROUP_ID: u8 = 0x0;
const GROUP_END: u8 = 0x03;

pub struct Group {
	pub subscribe_id: u64,
	pub track_alias: u64,
	pub group_id: u64,
	pub publisher_priority: u8,
}

impl Group {
	pub const STREAM_TYPE: u64 = 0x04;
}

impl Encode for Group {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.subscribe_id.encode(w);
		self.track_alias.encode(w);
		self.group_id.encode(w);
		SUBGROUP_ID.encode(w);
		self.publisher_priority.encode(w);
	}
}

impl Decode for Group {
	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let subscribe_id = u64::decode(r)?;
		let track_alias = u64::decode(r)?;
		let group_id = u64::decode(r)?;
		let subgroup_id = u8::decode(r)?;
		if subgroup_id != SUBGROUP_ID {
			return Err(DecodeError::InvalidValue);
		}
		let publisher_priority = u8::decode(r)?;
		Ok(Self {
			subscribe_id,
			track_alias,
			group_id,
			publisher_priority,
		})
	}
}

pub struct Frame {
	pub id: u64,
	pub payload: Option<Vec<u8>>,
}

impl Encode for Frame {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.id.encode(w);

		let size = self.payload.as_ref().map(|p| p.len()).unwrap_or(0);
		size.encode(w);

		match &self.payload {
			Some(payload) if !payload.is_empty() => w.put_slice(payload),
			Some(_) => 0u8.encode(w),
			None => GROUP_END.encode(w),
		}
	}
}

impl Decode for Frame {
	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let id = u64::decode(r)?;
		let size = u64::decode(r)?;

		if r.remaining() < size as usize {
			return Err(DecodeError::Short);
		}

		if size > 0 {
			let payload = r.copy_to_bytes(size as usize).to_vec();
			Ok(Self {
				id,
				payload: Some(payload),
			})
		} else {
			match u8::decode(r)? {
				0 => Ok(Self {
					id,
					payload: Some(Vec::new()),
				}),
				GROUP_END => Ok(Self { id, payload: None }),
				_ => Err(DecodeError::InvalidValue),
			}
		}
	}
}
